pkg/controller/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -527,6 +528,137 @@ func TestUpdateStatusToProvisionedWhenPreviouslyWasSetToFailed(t *testing.T) {
+		nsTmplSet := newNSTmplSet()
+		deletionTS := metav1.NewTime(time.Now())
+		nsTmplSet.SetDeletionTimestamp(&deletionTS) // mark resource as deleted
+		r, req, c := prepareReconcile(t, nsTmplSet)
+		for _, ns := range nsTmplSet.Spec.Namespaces {
+			createNamespace(t, r.client, ns.Revision, ns.Type)
+		}
+		c.MockDelete = func(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error {
+			if obj, ok := obj.(*corev1.Namespace); ok {
+				// mark namespaces as deleted...
+				deletionTS := metav1.NewTime(time.Now())
+				obj.SetDeletionTimestamp(&deletionTS)
+				// ... but replace them in the fake client cache yet instead of deleting them
+				return c.Client.Update(ctx, obj)
+			}
+			return c.Client.Delete(ctx, obj, opts...)
+		}
+
+		t.Run("reconcile after nstemplateset deletion", func(t *testing.T) {
+			// when a first reconcile loop is triggered (when the NSTemplateSet resource is marked for deletion and there's a finalizer)
+			_, err := r.Reconcile(req)
+
+			// then
+			require.NoError(t, err)
+			// get the first namespace and check its deletion timestamp
+			firstNS := corev1.Namespace{}
+			firstNSName := fmt.Sprintf("%s-%s", username, nsTmplSet.Spec.Namespaces[0].Type)
+			err = r.client.Get(context.TODO(), types.NamespacedName{
+				Name: firstNSName,
+			}, &firstNS)
+			require.NoError(t, err)
+			assert.NotNil(t, firstNS.GetDeletionTimestamp(), "expected a deletion timestamp on '%s' namespace", firstNSName)
+			// get the NSTemplateSet resource again and check its status
+			updateNSTemplateSet := toolchainv1alpha1.NSTemplateSet{}
+			err = r.client.Get(context.TODO(), types.NamespacedName{
+				Namespace: nsTmplSet.Namespace,
+				Name:      nsTmplSet.Name,
+			}, &updateNSTemplateSet)
+			require.NoError(t, err)
+			test.AssertConditionsMatch(t, updateNSTemplateSet.Status.Conditions, toolchainv1alpha1.Condition{
+				Type:   toolchainv1alpha1.ConditionReady,
+				Status: corev1.ConditionFalse,
+				Reason: toolchainv1alpha1.NSTemplateSetTerminatingReason,
+			})
+			// and the finalizer should NOt have been removed yet
+			assert.Equal(t, []string{toolchainv1alpha1.FinalizerName}, updateNSTemplateSet.Finalizers)
+
+			t.Run("reconcile after first user namespace deletion", func(t *testing.T) {
+				// given a second reconcile loop was triggered (because a user namespace was deleted)
+				_, req, _ := prepareReconcile(t, nsTmplSet)
+				// when
+				_, err := r.Reconcile(req)
+				// then
+				require.NoError(t, err)
+				// get the second namespace and check its deletion timestamp
+				secondNS := corev1.Namespace{}
+				secondtNSName := fmt.Sprintf("%s-%s", username, nsTmplSet.Spec.Namespaces[1].Type)
+				err = r.client.Get(context.TODO(), types.NamespacedName{
+					Name: secondtNSName,
+				}, &secondNS)
+				require.NoError(t, err)
+				assert.NotNil(t, secondNS.GetDeletionTimestamp(), "expected a deletion timestamp on '%s' namespace", secondtNSName)
+
```

Comment:
Let's also check the status, so it's still terminating. And the finalizer is still present.

Reply:
done in 1cc75b7